sdks/python/apache_beam/runners/portability/fn_api_runner/trigger_manager.py

```diff
@@ -0,0 +1,457 @@
+      finished_windows_state: SetRuntimeState = DoFn.StateParam(  # type: ignore
+          FINISHED_WINDOWS),
+      watermark_timer=DoFn.TimerParam(WATERMARK_TIMER)):
+    context = FnRunnerStatefulTriggerContext(
+        processing_time_timer=processing_time_timer,
+        watermark_timer=watermark_timer,
+        latest_processing_time=None,
+        latest_watermark=latest_watermark,
+        all_elements_state=all_elements,
+        window_tag_values=window_tag_values,
+        finished_windows_state=finished_windows_state)
+    result = self._fire_eligible_windows(
+        key, TimeDomain.WATERMARK, timestamp, timer_tag, context)
+    latest_watermark.add(timestamp)
+    return result
+
+
+class FnRunnerStatefulTriggerContext(TriggerContext):
+  def __init__(
+      self,
+      processing_time_timer: RuntimeTimer,
+      watermark_timer: RuntimeTimer,
+      latest_processing_time: typing.Optional[AccumulatingRuntimeState],
+      latest_watermark: typing.Optional[AccumulatingRuntimeState],
+      all_elements_state: BagRuntimeState,
+      window_tag_values: BagRuntimeState,
+      finished_windows_state: SetRuntimeState):
+    self.timers = {
+        TimeDomain.REAL_TIME: processing_time_timer,
+        TimeDomain.WATERMARK: watermark_timer
+    }
+    self.current_times = {
+        TimeDomain.REAL_TIME: latest_processing_time,
+        TimeDomain.WATERMARK: latest_watermark
+    }
+    self.all_elements_state = all_elements_state
+    self.window_tag_values = window_tag_values
+    self.finished_windows_state = finished_windows_state
+
+  def windows_to_elements_map(
+      self
+  ) -> typing.Dict[BoundedWindow, typing.List[windowed_value.WindowedValue]]:
+    window_element_pairs: typing.Iterable[typing.Tuple[
+        BoundedWindow,
+        windowed_value.WindowedValue]] = self.all_elements_state.read()
+    result: typing.Dict[BoundedWindow,
+                        typing.List[windowed_value.WindowedValue]] = {}
+    for w, e in window_element_pairs:
+      if w not in result:
+        result[w] = []
+      result[w].append(e)
+    return result
+
+  def for_window(self, window):
+    return PerWindowTriggerContext(window, self)
+
+  def get_current_time(self):
+    return self.current_times[TimeDomain.REAL_TIME].read()
+
+  def set_timer(self, name, time_domain, timestamp):
+    _LOGGER.debug('Setting timer (%s, %s) at %s', time_domain, name, timestamp)
+    self.timers[time_domain].set(timestamp, dynamic_timer_tag=name)
+
+  def clear_timer(self, name, time_domain):
+    _LOGGER.debug('Clearing timer (%s, %s)', time_domain, name)
+    self.timers[time_domain].clear(dynamic_timer_tag=name)
+
+  def merge_windows(self, to_be_merged, merge_result):
+    all_triplets = list(self.window_tag_values.read())
+    # Collect all the triplets for the window we are merging away, and tag them
+    # with the new window (merge_result).
+    merging_away_triplets = [(merge_result, t[1], t[2]) for t in all_triplets
```

Comment:
I think it will be more readable if we unpack the triplets for example 
```
[(merge_result, state_tag, state) for (window, state_tag, state) in all_triplets if window in to_be_merged]
```
WDYT? Same for other list comprehensions in this file.